src/Microsoft.Health.Fhir.Core/Features/Operations/Import/ImportOrchestratorTask.cs

```diff
@@ -0,0 +1,507 @@
+            {
+                TaskInfo completeTaskInfo = _orchestratorTaskContext.DataProcessingTasks[completedResourceUri];
+                TaskResultData taskResultData = JsonConvert.DeserializeObject<TaskResultData>(completeTaskInfo.Result);
+
+                if (taskResultData.Result == TaskResult.Success)
+                {
+                    ImportProcessingTaskResult procesingTaskResult = JsonConvert.DeserializeObject<ImportProcessingTaskResult>(taskResultData.ResultData);
+                    completedOperationOutcome.Add(new ImportOperationOutcome() { Type = procesingTaskResult.ResourceType, Count = procesingTaskResult.SucceedCount, InputUrl = completedResourceUri });
+                    if (procesingTaskResult.FailedCount > 0)
+                    {
+                        failedOperationOutcome.Add(new ImportFailedOperationOutcome() { Type = procesingTaskResult.ResourceType, Count = procesingTaskResult.FailedCount, InputUrl = completedResourceUri, Url = procesingTaskResult.ErrorLogLocation });
+                    }
+                }
+                else if (taskResultData.Result == TaskResult.Fail)
+                {
+                    throw new ImportProcessingException(string.Format("Failed to process file: {0}. {1}", completedResourceUri, taskResultData));
+                }
+                else if (taskResultData.Result == TaskResult.Canceled)
+                {
+                    throw new OperationCanceledException(taskResultData.ResultData);
+                }
+            }
+        }
+
+        private async Task<List<Uri>> MonitorRunningTasksAsync(List<(Uri resourceUri, TaskInfo taskInfo)> runningTasks, CancellationToken cancellationToken)
+        {
+            List<Uri> completedTaskResourceUris = new List<Uri>();
+
+            foreach ((Uri runningResourceUri, TaskInfo runningTaskInfo) in runningTasks)
+            {
+                if (cancellationToken.IsCancellationRequested)
+                {
+                    throw new OperationCanceledException();
+                }
+
+                TaskInfo latestTaskInfo = await _taskManager.GetTaskAsync(runningTaskInfo.TaskId, cancellationToken);
+
+                _orchestratorTaskContext.DataProcessingTasks[runningResourceUri] = latestTaskInfo;
+                if (latestTaskInfo.Status == TaskStatus.Completed)
+                {
+                    completedTaskResourceUris.Add(runningResourceUri);
+                }
+            }
+
+            return completedTaskResourceUris;
+        }
+
+        private async Task CancelProcessingTasksAsync()
+        {
+            List<string> runningTaskIds = new List<string>();
+
+            if ((_orchestratorTaskContext?.DataProcessingTasks?.Count ?? 0) == 0)
+            {
+                // No data processing task created.
+                return;
+            }
+
+            foreach (TaskInfo taskInfo in _orchestratorTaskContext.DataProcessingTasks.Values)
```

Comment:
Is it safer to check if it is null? _orchestratorTaskContext.DataProcessingTasks.Values

Reply:
We check it before using it.

            if ((_orchestratorTaskContext?.DataProcessingTasks?.Count ?? 0) == 0)
            {
                // No data processing task created.
                return;
            }